### ISSUE TITLE
Support for aggregate function with filter

### ISSUE THREAD

**Issue by jimexist** [apache/datafusion#272]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

A new challenge, to support aggregate function with filter. See [Postgres manual](https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-AGGREGATES)

**Describe the solution you'd like**

Exactly like [in the manual](https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-AGGREGATES)

**Describe alternatives you've considered**

**Additional context**

**Comment by Dandandan** [comment]
That would be a valuable feature @Jimexist

**Comment by alamb** [comment]
@Jimexist if you are proposing supporting this kind of syntax, then I think it is a reasonable idea. 

```
SELECT
    count(*) AS unfiltered,
    count(*) FILTER (WHERE i < 5) AS filtered
FROM generate_series(1,10) AS s(i);
 unfiltered | filtered
------------+----------
         10 |        4
(1 row)
```

For anyone else who might read this ticket, You can express the same type of query using SQL that data fusion already supports via CASE, though it is much less beautiful. Something like:

```shell
$ seq 1 10 > /tmp/foo.csv
$ cargo run -p datafusion-cli
```

```
> CREATE EXTERNAL TABLE t(i double) STORED AS CSV LOCATION '/tmp/foo.csv';

> select * from t;
+----+
| i  |
+----+
| 1  |
| 2  |
| 3  |
| 4  |
| 5  |
| 6  |
| 7  |
| 8  |
| 9  |
| 10 |
+----+

>    select count(*) as unfiltered, sum(case when i < 5 then 1 else 0 end) as filtered from t;
+------------+----------+
| unfiltered | filtered |
+------------+----------+
| 10         | 4        |
+------------+----------+
```

**Comment by Dandandan** [comment]
@alamb yes, that's a great one - isn't that only feasible with `SUM`/`MAX`/`MIN`? I think primary use case would be `COUNT` or `AVG`

**Comment by kawadakk** [comment]
This feature was implemented by #5868, although (afaik) you still need to select a SQL dialect supporting it.

**Comment by jayzhan211** [comment]
> This feature has been implemented by #5868. Note that (afaik) you still need to select a SQL dialect supporting it.

It seems we can close this issue
